Decouple linux process specifics from elf dump file output

dblnz opened the pull request:
## Objective
I would like to use the ELF core dump generation on a system that does not have linux kernel.
The ELF core dump generation should work also on windows.
I need to create a core dump from the information I have about the guest running in a hypervisor isolated [Hyperlight Sandbox](https://github.com/hyperlight-dev/hyperlight/blob/main/README.mdl) (registers and guest memory).

## Design decisions
- Maintain the existent public interface so that this does not introduce breaking changes
- Add the possibility to replace the Linux process as input for the ELF core dump creation
- Move the linux specific logic including the `ProcessView` type to a separate `linux` module that can be conditionally compiled when the platform is linux.
This was not needed, but I thought it would be easier to follow.
- Make some of the internal types public so that they can be used by the custom process info source (implies modifying visibility and add required documentation)

## Description
I defined a `trait` called `ProcessInfoSource` that would provide the necessary information the `elfcore` core dump output logic needs (pid, threads, regions, mapped_files, aux_vector, page_size).
The already defined `ProcessView` now implements this trait. The `elfcore` core dump output logic now takes a `ProcessInfoSource` trait object that provides all the needed information. This decouples the core dump generation from the `ProcessView` type that assumes a linux kernel.
To provide the custom process information source, the `from_source` method was added to `CoreDumpBuilder` to create a custom `CoreDumpBuilder` that takes the custom info from a provided source and memory reader.

chris-oo reviewed (commented):
The overall shape of the PR seems reasonable, but some minor nitpicks and questions about dead code.

dblnz commented on elfcore/src/linux.rs:
```diff
@@ -0,0 +1,745 @@
+// Copyright (c) Microsoft Corporation.
```
Yes, most of the contents were moved from `coredump.rs` and then modified to accept the trait object as input

dblnz commented on elfcore/src/lib.rs:
```diff
@@ -10,9 +10,32 @@ mod arch;
+/// By implementing this trait one can use the ELF output logic
+pub trait ProcessInfoSource {
+    /// Retrieves the PID of the process
+    fn get_pid(&self) -> nix::unistd::Pid;
```
I'll change the name of the methods and add more details to the documentation.

dblnz commented on elfcore/src/arch/mod.rs:
```diff
@@ -20,30 +20,39 @@ mod aarch64;
 pub use aarch64::elf_gregset_t;
 
 #[derive(Debug)]
+#[allow(dead_code)]
```
I remember getting a clippy warning about the `name` field not being used, but it looks like it was only the method from the `Arch` trait.
I'll remove the dead code from here

dblnz commented on elfcore/src/coredump.rs:
```diff
@@ -922,40 +250,34 @@ fn process_vm_readv_works() -> bool {
 fn write_core_dump_inner<T: Write>(
     writer: T,
-    pv: &ProcessView,
+    pv: &dyn ProcessInfoSource,
```
I agree, using generics should be fine. I'll make the changes.

dblnz replied:
I tried to change this function's signature to use generics, however it forced me to modify some public types which I do not believe would be desired. It would introduce a breaking change.
Let me know if you have an idea on how to do this without affecting the existing public api.

dblnz commented:
I took care of the comments above and modified the changes to address them and marked the conversation as resolved.
I modified the history of this branch to be easier to follow commit by commit.
I also added some tests and support for non linux platforms so we can generate ELF core dumps from `ProcessInfoSource` on any platforms.

dblnz commented:
I added a commit that changes the crate internally from using trait objects to generics.
The consequence is we have some breaking changes:
- the `CoreDumpBuilder` struct expects two generic arguments
- the `LinuxProcessMemoryReader` (internal type that handles linux process memory reading) needs to be public now
- to create a core dump for a linux process one needs to use `CoreDumpBuilder::<ProcessView, LinuxProcessMemoryReader>::new(pid)` instead of `CoreDumpBuilder::new(pid)`

chris-oo reviewed (commented):
other than the nested imports, LGTM. Happy to merge once that's fixed.

I assume you'll need a new release on crates.io? I'll see about also getting some folks on my team to review.

dblnz commented on elfcore-sample/src/main.rs:
```diff
@@ -11,16 +11,19 @@
-use std::path::PathBuf;
-use tracing::Level;
+#[cfg(target_os = "linux")]
+use {
```
Sure, I'll change them

dblnz commented:
Thanks for the review.
A new release on crates.io would be great.
